src/System.Management.Automation/engine/MshObject.cs

```diff
@@ -2347,6 +2378,204 @@ internal bool IsHelpObject
+            _validatedNames = null;
+            _currentObjectStack = null;
+            _currentObject = null;
+        }
+
+        /// <summary>
+        /// Begins the creation of a new PSObject
+        /// </summary>
+        /// <param name="capacity">The initial capacity for the members collection.</param>
+        public void BeginCreateObject(int capacity)
+        {
+            if (_currentObject != null)
+            {
+                CurrentObjectStack.Push(_currentObject);
+            }
+            _currentObject = new PSObject(capacity);
+        }
+
+        /// <summary>
+        /// Begins the creation of a new PSObject
+        /// </summary>
+        /// <param name="instance">The object to wrap in a PSObject.</param>
+        /// <param name="capacity">The initial capacity for the members collection.</param>
+        public void BeginCreateObject(object instance, int capacity)
+        {
+            if (_currentObject != null)
+            {
+                CurrentObjectStack.Push(_currentObject);
+            }
+            _currentObject = new PSObject(instance, capacity);
+        }
+
+
+        /// <summary>
+        /// Add a property to the object.
+        /// </summary>
+        /// <param name="name">The name of the property.</param>
+        /// <param name="value">The value of the property.</param>
+        public void AddNoteProperty(string name, object value)
+        {
+            var validated = _validatedNames?.Contains(name) ?? true;
+            if (!validated)
+            {
+                _validatedNames.Add(name);
+            }
+            _currentObject.Properties.Add(new PSNoteProperty(name, value), validated, inObjectBuilderScope:true);
+        }
+
+        /// <summary>
+        /// Completes the addition of the object.
+        /// </summary>
+        /// <returns></returns>
+        public PSObject EndCreateObject()
+        {
+            var current = _currentObject;
+            PSObject next = null;
+            CurrentObjectStack?.TryPop(out next);
+            _currentObject = next;
+            return current;
+        }
+    }
+
+    /// <summary>
+    /// Unsafe version of PSObjectBuilder that is not guaranteed to be on the same thread.
+    /// Usage must be reviewed carefully.
+    /// </summary>
+    internal class PSObjectBuilderUnsafe
+    {
+        const int DefaultHirearchyDepth = 4;
+        private readonly HashSet<string> _validatedNames;
+        private Stack<PSObject> _currentObjectStack;
+
+        private PSObject _currentObject;
+
+        private Stack<PSObject> CurrentObjectStack => _currentObjectStack ?? (_currentObjectStack = new Stack<PSObject>(DefaultHirearchyDepth));
+
+        /// <summary>
+        /// Initializes a new instance of PSObjectBuilder.
+        /// </summary>
+        /// <param name="validatedNameCount">The initial size of the set that tracks validated parameter names.</param>
+        public PSObjectBuilderUnsafe(int validatedNameCount)
```

Comment:
It should be internal only if it is unsafe. On the other hand, it would be nice to have an api that would work in scripts not only in binaries.

Reply:
I think we should have another name. I actually don't mean unsafe, I mean slower since we cannot make as many guarantees as the ref struct.

Reply:
So it's a really bad name.